server/app/scanpy_engine/scanpy_engine.py

```diff
@@ -214,51 +225,53 @@ def layout(self, df):
-        mean_ex1_set1 = np.mean(expression_1[:, set1], axis=0)[sort_set1]
-        mean_ex2_set1 = np.mean(expression_2[:, set1], axis=0)[sort_set1]
-        mean_ex1_set2 = np.mean(expression_1[:, set2], axis=0)[sort_set2]
-        mean_ex2_set2 = np.mean(expression_2[:, set2], axis=0)[sort_set2]
-        mean_diff1 = mean_ex1_set1 - mean_ex2_set1
-        mean_diff2 = mean_ex1_set2 - mean_ex2_set2
-        genes_cellset_1 = self.data.var_names[set1][sort_set1]
-        genes_cellset_2 = self.data.var_names[set2][sort_set2]
-        return {
-            "celllist1": {
-                "topgenes": genes_cellset_1.tolist()[:num_genes],
-                "mean_expression_cellset1": mean_ex1_set1.tolist()[:num_genes],
-                "mean_expression_cellset2": mean_ex2_set1.tolist()[:num_genes],
-                "pval": pval1.tolist()[:num_genes],
-                "ave_diff": mean_diff1.tolist()[:num_genes]
-            },
-            "celllist2": {
-                "topgenes": genes_cellset_2.tolist()[:num_genes],
-                "mean_expression_cellset1": mean_ex1_set2.tolist()[:num_genes],
-                "mean_expression_cellset2": mean_ex2_set2.tolist()[:num_genes],
-                "pval": pval2.tolist()[:num_genes],
-                "ave_diff": mean_diff2.tolist()[:num_genes]
-            },
-        }
+        # If not the same genes, test is wrong!
+        if np.any(df1.var.index != df2.var.index):
+            raise ValueError("Variables ares not the same in set1 and set2")
+
+        # If not all genes, they used a var filter
+        if df1.var.shape[0] < self.gene_count:
+            mode = DiffExpMode("var filter")
+            if top_n:
+                raise Warning("Top N was specified but will not be used in 'Var Filter' mode")
+        else:
+            mode = DiffExpMode("top var")
+            if not top_n:
+                top_n = DEFAULT_TOP_N
+
+        genes_idx = df1.var.index
+        diffexp_result = stats.ttest_ind(df1.X, df2.X)
+        pval = diffexp_result.pvalue
+        bonferroni_pval = 1 - (1 - pval) ** self.gene_count
+        ave_exp_set1 = np.mean(df1.X, axis=0)
+        ave_exp_set2 = np.mean(df2.X, axis=0)
+        ave_diff = ave_exp_set1 - ave_exp_set2
+        if mode == DiffExpMode.TOP_VAR:
+            sort_order = np.argsort(np.abs(diffexp_result.statistic))[::-1]
+            # If top_n > length it will just return length
+            genes = self._top_sort(genes_idx, sort_order, top_n)
+            pval = self._top_sort(pval, sort_order, top_n)
+            bonferroni_pval = self._top_sort(bonferroni_pval, sort_order, top_n)
+            ave_exp_set1 = self._top_sort(ave_exp_set1, sort_order, top_n)
+            ave_exp_set2 = self._top_sort(ave_exp_set2, sort_order, top_n)
+            ave_diff = self._top_sort(ave_diff, sort_order, top_n)
+
+        # varIndex, avgDiff,  pVal, pValAdj, set1AvgExp, set2AvgExp
+        result = []
+        for i in range(len(genes)):
```

Comment:
I don't know if `zip(genes, ave_diff, pval, ...)` would work or even be better, but maybe...?